jormungandr/src/fragment/pool.rs

```diff
@@ -192,59 +190,40 @@ pub(super) mod internal {
-                self.expirations.remove(&cache_key);
-                Some(fragment)
-            } else {
-                None
-            }
-        }
-
         pub fn remove_all(&mut self, fragment_ids: impl IntoIterator<Item = FragmentId>) {
-            // TODO fix terrible performance, entries_by_time are linear searched N times
-            for fragment_id in fragment_ids {
-                self.remove(&fragment_id);
+            let to_remove = fragment_ids.into_iter().collect::<HashSet<_>>();
+            for fragment_id in to_remove.iter() {
+                if let Some(PoolEntryInternal { expiration_key, .. }) =
+                    self.entries.remove(fragment_id)
+                {
+                    self.expirations.remove(expiration_key);
+                }
+            }
+            let mut clean_entries_by_time = VecDeque::new();
+            for entry in self.entries_by_time.iter().cloned() {
+                if !to_remove.contains(&entry) {
+                    clean_entries_by_time.push_back(entry);
+                }
             }
+            self.entries_by_time = clean_entries_by_time;
         }
 
         pub fn remove_oldest(&mut self) -> Option<Fragment> {
             let fragment_id = self.entries_by_time.pop_front()?;
-            let (_, fragment, cache_key) = self
+            let PoolEntryInternal {
+                fragment,
+                expiration_key,
+            } = self
                 .entries
                 .remove(&fragment_id)
                 .expect("Pool lost fragment ID consistency");
-            self.expirations.remove(&cache_key);
+            self.expirations.remove(expiration_key);
             Some(fragment)
         }
 
-        pub fn poll_purge(&mut self, cx: &mut Context) -> Poll<Result<(), time::Error>> {
-            loop {
-                match self.expirations.as_mut().poll_next(cx) {
-                    Poll::Ready(Some(Ok(entry))) => {
-                        self.entries.remove(entry.get_ref());
-                        self.entries_by_time
-                            .iter()
-                            .position(|id| id == entry.get_ref())
-                            .map(|position| {
-                                self.entries_by_time.remove(position);
-                            });
-                    }
-                    Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
-                    Poll::Ready(None) => return Poll::Ready(Ok(())),
-
-                    // Here Pending means there are still items in the DelayQueue but
-                    // they are not expired. We don't want this function to wait for these
-                    // ones to expired. We only cared about removing the expired ones.
-                    Poll::Pending => return Poll::Ready(Ok(())),
-                }
-            }
+        pub fn purge(&mut self) {
+            let expired = self.expirations.pop_expired();
+            self.remove_all(expired);
```

Comment:
I wonder if making `pop_expired` return an `impl Iterator` and then piping that into `.remove_all` would make it more efficient. But this can wait for another PR.